src/net/send_recv/msg.rs

```diff
@@ -0,0 +1,662 @@
+            let last_header = last_header.as_mut();
+            last_header.cmsg_len = c::CMSG_LEN(source_len) as _;
+            last_header.cmsg_level = cmsg_level;
+            last_header.cmsg_type = cmsg_type;
+        }
+
+        // Get the pointer to the payload and copy the data.
+        unsafe {
+            let payload = c::CMSG_DATA(last_header.as_ptr());
+            core::ptr::copy_nonoverlapping(source.as_ptr(), payload, source_len as _);
+        }
+
+        true
+    }
+}
+
+impl<'slice, 'fd> Extend<SendAncillaryMessage<'slice, 'fd>>
+    for SendAncillaryBuffer<'_, 'slice, 'fd>
+{
+    fn extend<T: IntoIterator<Item = SendAncillaryMessage<'slice, 'fd>>>(&mut self, iter: T) {
+        // TODO: This could be optimized to add every message in one go.
+        iter.into_iter().all(|msg| self.push(msg));
+    }
+}
+
+/// Buffer for receiving ancillary messages.
+pub struct RecvAncillaryBuffer<'buf> {
+    /// Raw byte buffer for messages.
+    buffer: &'buf mut [u8],
+
+    /// The portion of the buffer we've read from already.
+    read: usize,
+
+    /// The amount of the buffer that is used.
+    length: usize,
+}
+
+impl<'buf> From<&'buf mut [u8]> for RecvAncillaryBuffer<'buf> {
+    fn from(buffer: &'buf mut [u8]) -> Self {
+        Self::new(buffer)
+    }
+}
+
+impl Default for RecvAncillaryBuffer<'_> {
+    fn default() -> Self {
+        Self::new(&mut [])
+    }
+}
+
+impl<'buf> RecvAncillaryBuffer<'buf> {
+    /// Create a new, empty `RecvAncillaryBuffer` from a raw byte buffer.
+    pub fn new(buffer: &'buf mut [u8]) -> Self {
+        Self {
+            buffer,
+            read: 0,
+            length: 0,
+        }
+    }
+
+    /// Returns a pointer to the message data.
+    pub(crate) fn as_control_ptr(&mut self) -> *mut u8 {
+        self.buffer.as_mut_ptr()
+    }
+
+    /// Returns the length of the message data.
+    pub(crate) fn control_len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Drain all messages from the buffer.
+    #[allow(unsafe_code)]
+    pub fn drain(&mut self) -> AncillaryDrain<'_> {
+        AncillaryDrain {
+            messages: messages::Messages::new(
+                &mut self.buffer[self.read..self.read + self.length],
+                self.length,
+            ),
+            read: &mut self.read,
+        }
+    }
+}
+
+impl Drop for RecvAncillaryBuffer<'_> {
+    fn drop(&mut self) {
+        self.drain().for_each(drop);
+    }
+}
+
+/// An iterator that drains messages from a `RecvAncillaryBuffer`.
+pub struct AncillaryDrain<'buf> {
+    /// Inner iterator over messages.
+    messages: messages::Messages<'buf>,
+
+    /// Increment the number of messages we've read.
+    read: &'buf mut usize,
+}
+
+impl<'buf> AncillaryDrain<'buf> {
+    /// A closure that converts a message into a `RecvAncillaryMessage`.
+    #[allow(unsafe_code)]
+    fn cvt_msg(read: &mut usize, msg: NonNull<c::cmsghdr>) -> Option<RecvAncillaryMessage<'buf>> {
+        unsafe {
```

Comment:
Ordinarily, a function that takes a raw pointer like this should be an `unsafe` function. But since this is not public, and it would add a fair amount of clutter to make it unsafe, I think it's ok it is. But it would be good to have more comments about what unsafe things are being done here.

Reply:
Not needed after I changed it to `&c::msghdr`.